src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp

```diff
@@ -860,7 +786,9 @@ ShenandoahConcurrentEvacThreadClosure::ShenandoahConcurrentEvacThreadClosure(Oop
 void ShenandoahConcurrentEvacThreadClosure::do_thread(Thread* thread) {
   JavaThread* const jt = JavaThread::cast(thread);
   StackWatermarkSet::finish_processing(jt, _oops, StackWatermarkKind::gc);
-  ShenandoahThreadLocalData::enable_plab_promotions(thread);
+  if (ShenandoahHeap::heap()->mode()->is_generational()) {
```

Comment:
Yes, there are many cases where the singleton is accessed directly. I also do not prefer this approach because it makes it difficult to run such classes in a unit test. I will refactor this.